Lib/tarfile.py

```diff
@@ -1520,12 +1526,26 @@ def _proc_pax(self, tarfile):
     def _proc_gnusparse_00(self, next, pax_headers, buf):
         """Process a GNU tar extended sparse header, version 0.0.
         """
-        offsets = []
-        for match in re.finditer(br"\d+ GNU.sparse.offset=(\d+)\n", buf):
-            offsets.append(int(match.group(1)))
-        numbytes = []
-        for match in re.finditer(br"\d+ GNU.sparse.numbytes=(\d+)\n", buf):
-            numbytes.append(int(match.group(1)))
+        def finditer_without_backtracking(buf, needle):
+            values = []
+            regex = re.compile(br"^\d+%s(\d{1,20})\n" % (needle,))
+            while True:
+                if (
+                    # Statement is both a contains check (!=-1) and a bounds check (>0)
+                    (needle_offset := buf.find(needle) - 1) > -1
+                    # Check that the character before is a digit (0x30-0x39 is 0-9)
+                    and 0x30 <= buf[needle_offset] <= 0x39
+                ):
+                    match = regex.match(buf[needle_offset:])
+                    if match is not None:
+                        values.append(int(match.group(1)))
+                    buf = buf[needle_offset + len(needle):]  # Skip over to the match.
+                else:
+                    break
+            return values
+
+        offsets = finditer_without_backtracking(buf, b" GNU.sparse.offset=")
+        numbytes = finditer_without_backtracking(buf, b" GNU.sparse.numbytes=")
```

Comment:
Just use `br"\d GNU.sparse.offset=(\d+)\n"` or `br"\d GNU.sparse.offset=(\d++)\n"`.

Reply:
I suppose this is similar to above where the ignored probable length prefix doesn't need to be anchored and have its full digit values matched?  that makes this a bit more lenient than it already is such that it could possibly match things it should not in unobvious situations.  if the \d+ length prefix happens to be checked elsewhere as appears to be the case if this is only called by the earlier `._proc_pax` code that does that maybe it matters less. But what if someone encoded a value that matches these patterns but wasn't actually a header?  this blind regex approach already believes them even though they were not intended as a header and possibly malicious so that different software parsed them in a different manner...

Reply:
It matches the same things that before, but without backtracking.

Reply:
I changed the implementation to use already-parsed headers: https://github.com/python/cpython/pull/121286/commits/4d04f3bc24f2f3e679722f8a06dee801bcff64ed Let me know what you think of this implementation.